# 表示质量比转摩尔比   1表示摩尔比转质量比

In [31]:
import re

In [2]:
# 0表示质量 比转摩尔比   1表示摩尔 比转质量比
tran_type = 0
# 元素类型及占比

proportion = {
    'SiO2':49.31
,   'Na2O':21.21
,   'B2O3':  5.88
,   'Al2O3':22.3
,   'NiO':1.31
}

In [19]:
# 自己创建一个常见元素类型的字典 以来查询质量
elements = { 'Li':6.941, 'Be':9.012182, 'B':10.811, 'C':12.0107, 'N':14.0067, 'O':15.9994, 'F':18.9984032, 'Na':22.98976, 'Mg':24.3050,
            'Al': 26.9815,  'Si':28.0855,'P':30.97376, 'S':32.065,'Cl':35.453,'K':39.098, 'Ca':40.078,'Sc':44.955912,'Ti':47.867, 
            'V':50.9415,'Cr':51.9961,'Mn':54.938045,'Fe':55.845,'Co':58.933195,'Ni':58.693,'Cu':63.546,'Zn':65.38,'Ga':69.723,
            'Ge':72.64,'Nb':92.9063,'As':74.92160,'Se':78.96,'Br':79.904,'Kr':83.798,'Rb':85.4678,'Sr':87.62,'Y':88.90585,'Zr':91.224,
            'Mo':95.94,'Tc':97.9072,'Ru':101.07,'Rh':102.90550,'Ag':107.8682,'Cd':112.411,'In':14.818,'Sn':118.710,'Sb':121.760,
            'Te':127.60,'I':126.90447,'Cs':132.9054519,'Ba':137.327,'La':138.90547,'Ce':140.116,'Pr':140.90765,'Nd':144.242,'Bi':208.98040,
           'H':1.00794,}


In [101]:
# 来一个根据分子名称输出相对分子质量的程序
def a_get_molecular_mass(a_molecular_name):
    a_name_split = re.findall('[A-Z][^A-Z]*',a_molecular_name)
    a_molecular_mass = 0
    for a_i in a_name_split:
        if 'a'<= a_i[-1] <= 'z' or 'A' <= a_i[-1] <='Z':
            #print('元素',a_i,'有一个')
            a_molecular_mass += elements[a_i]
        else:
            #print('元素',a_i[:-1],'有',a_i[-1],'个')
            a_molecular_mass += elements[a_i[:-1]] * float(a_i[-1])
    return a_molecular_mass

In [114]:
xxx = 'Na2CO3'
a_get_molecular_mass(xxx)

105.98841999999999

In [99]:
# 把分子式切割
Sample_name = 'Li2CO3,NaPO3,KPO3,Rb2CO3,Cs2CO3,MgO,SrCO3,Ba(PO3)2,\
La2O3,Ce(NO3)3.6H2O,Nd2O3,ZrO2,V2O5,Cr2O3,MoO3,MnO2,Fe2O3,Co2O3,NiO,H3BO3,\
Al(H2PO4)3,SiO2,NH4H2PO4,Sb2O3,Bi2O3,TeO2,BaCO3,Al(OH)3'

In [100]:
Sample_name.split(',')

['Li2CO3',
 'NaPO3',
 'KPO3',
 'Rb2CO3',
 'Cs2CO3',
 'MgO',
 'SrCO3',
 'Ba(PO3)2',
 'La2O3',
 'Ce(NO3)3.6H2O',
 'Nd2O3',
 'ZrO2',
 'V2O5',
 'Cr2O3',
 'MoO3',
 'MnO2',
 'Fe2O3',
 'Co2O3',
 'NiO',
 'H3BO3',
 'Al(H2PO4)3',
 'SiO2',
 'NH4H2PO4',
 'Sb2O3',
 'Bi2O3',
 'TeO2',
 'BaCO3',
 'Al(OH)3']

In [104]:
get_molecular_mass('Al(OH)3')

元素 Al 有 ( 个
元素 O 有一个
元素 H) 有 3 个


In [65]:
elements['Si']

28.0855

In [4]:
# 化合物的分子摩尔质量
compound ={'SiO2':elements['Si']+elements['O']*2
          ,'Na2O':elements['Na']*2+elements['O']
          ,'B2O3':elements['B']*2+elements['O']*3
          ,'Al2O3':elements['Al']*2+elements['O']*3
          ,'NiO':elements['Ni']+ elements['O']
          }

In [87]:
compound['SiO2']

60.0843

# 主程序

In [3]:
if tran_type==0:
    print("得出摩尔的百分比：")
    mol_dic = {}
    for i in proportion:
        mol_dic[i]=proportion[i]/compound[i]
    print(mol_dic)
    sum_all = 0
    for i in mol_dic:
        sum_all+=mol_dic[i]
    for i in mol_dic:
        mol_dic[i]= round(mol_dic[i]/sum_all*100,2)
        print(i,' : ',mol_dic[i])
else:
    print("得出质量的百分比：")
    mass_dic = {}
    for i in proportion:
        mass_dic[i]=proportion[i] * compound[i]
    sum_all = 0
    for i in mass_dic:
        sum_all += mass_dic[i]
    for i in mass_dic:
        mass_dic[i]= round(mass_dic[i]/sum_all*100,2)
        print(i,' : ',mass_dic[i])

得出摩尔的百分比：
{'SiO2': 0.8206802775433849, 'Na2O': 0.34221312665661163, 'B2O3': 0.08445824631356992, 'Al2O3': 0.2187106467950554, 'NiO': 0.01753859830451291}
SiO2  :  55.32
Na2O  :  23.07
B2O3  :  5.69
Al2O3  :  14.74
NiO  :  1.18


mol
'SiO2':55.32
,   'Na2O':23.07
,   'B2O3': 5.69
,   'Al2O3':14.74
,   'NiO':1.18

得出质量的百分比：
SiO2  :  49.31
Na2O  :  21.21
B2O3  :  5.88
Al2O3  :  22.3
NiO  :  1.31